aws/resource_aws_lambda_layer.go

```diff
@@ -0,0 +1,262 @@
+				Computed: true,
+			},
+			"created_date": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"source_code_hash": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Computed: true,
+			},
+			"source_code_size": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"version": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsLambdaLayerPublish(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).lambdaconn
+
+	layerName := d.Get("layer_name").(string)
+	filename, hasFilename := d.GetOk("filename")
+	s3Bucket, bucketOk := d.GetOk("s3_bucket")
+	s3Key, keyOk := d.GetOk("s3_key")
+	s3ObjectVersion, versionOk := d.GetOk("s3_object_version")
+
+	if !hasFilename && !bucketOk && !keyOk && !versionOk {
+		return errors.New("filename or s3_* attributes must be set")
+	}
+
+	var layerContent *lambda.LayerVersionContentInput
+	if hasFilename {
+		awsMutexKV.Lock(awsMutexLambdaLayerKey)
+		defer awsMutexKV.Unlock(awsMutexLambdaLayerKey)
+		file, err := loadFileContent(filename.(string))
+		if err != nil {
+			return fmt.Errorf("Unable to load %q: %s", filename.(string), err)
+		}
+		layerContent = &lambda.LayerVersionContentInput{
+			ZipFile: file,
+		}
+	} else {
+		if !bucketOk || !keyOk {
+			return errors.New("s3_bucket and s3_key must all be set while using s3 code source")
+		}
+		layerContent = &lambda.LayerVersionContentInput{
+			S3Bucket: aws.String(s3Bucket.(string)),
+			S3Key:    aws.String(s3Key.(string)),
+		}
+		if versionOk {
+			layerContent.S3ObjectVersion = aws.String(s3ObjectVersion.(string))
+		}
+	}
+
+	params := &lambda.PublishLayerVersionInput{
+		Content:     layerContent,
+		Description: aws.String(d.Get("description").(string)),
+		LayerName:   aws.String(layerName),
+		LicenseInfo: aws.String(d.Get("license_info").(string)),
+	}
+
+	if v, ok := d.GetOk("compatible_runtimes"); ok && v.(*schema.Set).Len() > 0 {
+		params.CompatibleRuntimes = expandStringList(v.(*schema.Set).List())
+	}
+
+	log.Printf("[DEBUG] Publishing Lambda layer: %s", params)
+	result, err := conn.PublishLayerVersion(params)
+	if err != nil {
+		return fmt.Errorf("Error creating lambda layer: %s", err)
+	}
+
+	d.SetId(aws.StringValue(result.LayerVersionArn))
+	return resourceAwsLambdaLayerRead(d, meta)
+}
+
+func resourceAwsLambdaLayerRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).lambdaconn
+
+	layerName, version, err := resourceAwsLambdaLayerParseId(d.Id())
+	if err != nil {
+		return fmt.Errorf("Error parsing lambda layer ID: %s", err)
+	}
+
+	layerVersion, err := conn.GetLayerVersion(&lambda.GetLayerVersionInput{
+		LayerName:     aws.String(layerName),
+		VersionNumber: aws.Int64(version),
+	})
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "ResourceNotFoundException" && !d.IsNewResource() {
+			d.SetId("")
+			return nil
+		}
+		return err
```

Comment:
Nit: We should provide context about when this error occurred for operators and code maintainers, e.g.

```suggestion
		return fmt.Errorf("error reading Lambda Layer Version (%s): %s", d.Id(), err)
```